pkg/trtl/replica/sync.go

```diff
@@ -0,0 +1,491 @@
+func (r *Service) initiatorPhase2(ctx context.Context, wg *sync.WaitGroup, log zerolog.Logger, sender *streamSender, stream gossipStream) {
+	// Ensure that this routine signals when it exits
+	defer wg.Done()
+	log.Trace().Msg("starting initiator phase 2")
+
+	// Ensure that we close the sending channel when this routine exits to prevent
+	// deadlocks if this phase ends prematurely (e.g. the timeout expires).
+	defer sender.Close()
+
+	// Track how many check versions, updates, and repairs we get in phase 2
+	var versions, updates, repairs uint64
+	var err error
+
+gossip:
+	for {
+		// Check to make sure the deadline isn't over
+		select {
+		case <-ctx.Done():
+			log.Debug().Msg("context canceled while trying to recv messages from remote")
+			return
+		default:
+		}
+
+		// Read the next message from the remote replica
+		var sync *replica.Sync
+		if sync, err = stream.Recv(); err != nil {
+			if err != io.EOF {
+				// If the error is not EOF then something has gone wrong.
+				log.Error().Err(err).Msg("anti-entropy aborted early with recv error")
+			}
+			return
+		}
+
+		switch sync.Status {
+		case replica.Sync_CHECK:
+			// Check to see if this replica's version is later than the remote's, if
+			// so, send our later version back to that replica.
+			versions++
+			var local *object.Object
+			if local, err = r.db.Object(sync.Object.Key, options.WithNamespace(sync.Object.Namespace)); err != nil {
+				log.Warn().Err(err).
+					Str("namespace", sync.Object.Namespace).
+					Str("key", b64e(sync.Object.Key)).
+					Msg("failed check sync on initiator: could not fetch object meta")
+
+				// Even if the error is not found, since this routine is the
+				// initiator, it will return an error for a check request because
+				// we do not want to bounce check requests back and forth. E.g.
+				// check requests only initiate from the "pull" routine (phase 1
+				// locally, and phase 2 on the remote).
+				sync.Status = replica.Sync_ERROR
+				sync.Error = err.Error()
+				sender.Send(sync)
+				continue gossip
+			}
+
+			// Check if the local version is later; because this is the initiating
+			// routine, do nothing if the remote is a later version.
+			if local.Version.IsLater(sync.Object.Version) {
+				// Send the local object back to the remote
+				sync.Status = replica.Sync_REPAIR
+				sync.Object = local
+				sync.Error = ""
+				if ok := sender.Send(sync); ok {
+					updates++
+				}
+			}
+
+		case replica.Sync_REPAIR:
+			// Receiving a repaired object, check if it's still later than our local
+			// replica's and if so, save it to disk with an update instead of a put.
+			//
+			// NOTE: we must check if it's later than our local version in case a
+			// concurrent Put (a stomp) or concurrent synchronization updated it.
+			// NOTE: honu.Update performs the version checking in a transaction.
+			if err = r.db.Update(sync.Object, options.WithNamespace(sync.Object.Namespace)); err != nil {
```

Comment:
Is this the right place to track a skip or stomp?

Reply:
It is - and now that we have the update type functionality, I've added a TODO here for that tracking.